src/rebalancers/uniswap-v3/UniswapV3Rebalancer.sol

```diff
@@ -0,0 +1,478 @@
+        uint256 id = assetData.assetIds[0];
+
+        // Fetch and cache all position related data.
+        PositionState memory position = getPositionState(id, lowerTick, upperTick, initiator);
+
+        // Check that pool is initially balanced.
+        // Prevents sandwiching attacks when swapping and/or adding liquidity.
+        if (isPoolUnbalanced(position)) revert UnbalancedPool();
+
+        {
+            uint256 totalLiquidity = uint256(IUniswapV3Pool(position.pool).liquidity());
+            uint256 maxLiquidity = totalLiquidity.mulDivDown(LIQUIDITY_TRESHOLD, 1e18);
+            if (position.liquidity > maxLiquidity) revert LiquidityTresholdExceeded();
+        }
+
+        // Rebalance the position so that the maximum liquidity can be added for new ticks.
+        // The Pool must still be balanced after the swap.
+        (bool zeroToOne, uint256 amountOut, int24 tickChange) = getSwapParameters(position, id);
+        position.newLowerTick += tickChange;
+        position.newUpperTick += tickChange;
+        if (_swap(position, zeroToOne, amountOut)) revert UnbalancedPool();
+
+        // Increase liquidity of the position.
+        // The approval for at least one token after increasing liquidity will remain non-zero.
+        // We have to set approval first to 0 for ERC20 tokens that require the approval to be set to zero
+        // before setting it to a non-zero value.
+        uint256 balance0 = ERC20(position.token0).balanceOf(address(this));
+        uint256 balance1 = ERC20(position.token1).balanceOf(address(this));
+        ERC20(position.token0).safeApprove(address(UniswapV3Logic.POSITION_MANAGER), 0);
+        ERC20(position.token0).safeApprove(address(UniswapV3Logic.POSITION_MANAGER), balance0);
+        ERC20(position.token1).safeApprove(address(UniswapV3Logic.POSITION_MANAGER), 0);
+        ERC20(position.token1).safeApprove(address(UniswapV3Logic.POSITION_MANAGER), balance1);
+
+        (uint256 newTokenId,, uint256 amount0, uint256 amount1) = UniswapV3Logic.POSITION_MANAGER.mint(
+            MintParams({
+                token0: position.token0,
+                token1: position.token1,
+                fee: position.fee,
+                tickLower: position.newLowerTick,
+                tickUpper: position.newUpperTick,
+                amount0Desired: balance0,
+                amount1Desired: balance1,
+                amount0Min: 0,
+                amount1Min: 0,
+                // TODO: send direct to actionHandler ?
+                recipient: address(this),
+                deadline: block.timestamp
+            })
+        );
+
+        // Update the tokenId for newly minted position
+        assetData.assetIds[0] = newTokenId;
+
+        // Any surplus is transferred to the Account.
+        balance0 -= amount0;
+        balance1 -= amount1;
+        if (balance0 > 0) ERC20(position.token0).safeTransfer(account_, balance0);
```

Comment:
Can this be deposited again in the account as well (ie added to assetData and approved for the account)?

Reply:
the newly minted position you mean ? The AM is not the owner of the Account, so no direct deposit call possible no ?

Reply:
It is a flashaction -> the assetdata this function returns is to deposit.
Both the position as remaining assets must be back deposited.

Reply:
done